starlite/config/logging.py

```diff
@@ -9,6 +19,137 @@
 except ImportError:
     picologging_config = None
 
+from starlite.connection import Request
+from starlite.enums import ScopeType
+from starlite.middleware.base import DefineMiddleware, MiddlewareProtocol
+
+if TYPE_CHECKING:
+    from starlette.types import ASGIApp, Message, Receive, Scope, Send
+
+
+class LoggingMiddleware(MiddlewareProtocol):
+    def __init__(
+        self,
+        app: "ASGIApp",
+        config: "LoggingConfig",  # pylint: disable=redefined-outer-name
+    ):
+        """Logging Middleware class.
+
+        This Middleware log incoming request and outgoing response. It use classic logging library.
+
+        Examples:
+
+            ```python
+            from starlite import Starlite, Request, get
+            from starlite.config.logging import LoggingConfig
+
+            logging_config = LoggingConfig(middleware_log_request=False)
+
+
+            @get("/")
+            def my_handler(request: Request) -> None:
+                ...
+
+
+            app = Starlite(route_handlers=[my_handler], middleware=[logging_config.middleware])
+            ```
+
+        You can use another logging library by subclassing LoggingMiddleware.
+
+        Examples:
+
+            ```python
+            from starlite import Starlite, Request, get
+            from starlite.config.logging import LoggingMiddleware, LoggingConfig
+
+
+            class PicologgingMiddleware(LoggingMiddleware):
+                def get_logger(self):
+                    import picologging as logging
+
+                    return logging.getLogger("custom.path.to.logging")
+
+
+            logging_config = LoggingConfig(
+                middleware_log_request=False, middleware_class=PicologgingMiddleware
+            )
+
+
+            @get("/")
+            def my_handler(request: Request) -> None:
+                ...
+
+
+            app = Starlite(route_handlers=[my_handler], middleware=[logging_config.middleware])
+            ```
+
+        Args:
+            app: The 'next' ASGI app to call.
+            config: The LoggingConfig instance.
+        """
+        super().__init__(app)
+        self.app = app
+        self.config = config
+        self.logger = self.get_logger()
+
+    def get_logger(self) -> Logger:
```

Comment:
Why don't we use the same method in configure to return the logger for picologging if it's in the class config?

Reply:
See documentation, I recommand to subclass for adding picologging support. We can provide a PicologgingMiddleware if you'd like. Moreover, we could move picologging configuration very strange piece of code in here.

Reply:
My only hesitation is that it now requires additional middleware instead of a logging config change in order to get `picologging` working.  Given that there are a few outstanding bugs right now with full `picologging` integration, I'm inclined to ignore it for now.  

However, in general, I think that we should focus on making it simple to configure.  In my mind, that means the logging middleware should automatically use `picologging` if it's configured in the LoggingConfig, but I could be persuaded otherwise.

@Goldziher  @peterschutt  you guys have strong feelings either way?

Reply:
I think the Middleware should be agnostic to the logging config. The question is how to achieve this. 

Optimally the Middleware would be able to call a function and receive the configured logger - this could be stdlib, picologger or anything else. 

Tell me, can we use `logging.getLogger` with picologging? If so I would just use this and decouple the Middleware from the logging config. They don't need to be tightly coupled.

Reply:
Yeah, I think we can.  If i remember correctly in the tests, we show that get logger retuned from `logging.getLogger` is the picologger when it is configured.

Reply:
Even if API are identical between both logging and picologging, we need to import the correct one and call the selected one when initializing it and when we try to fetch a logger in a standard part of starlite or in a third party.

I see tow possibilities:
1/ add a config option (something like logging_class) and check that option when calling logging.config.dictConfig()
2/ add a subclass of LoggingConfig called PicologgingConfig (which is almost the same as before)

I would like to avoid any configuration parsing to find a keyword. It looks like a needle in a haystack, it's hard to explain and seems to me very prone to unwanted behavior.

To untight Middleware from config, we  need to provide a straightforward way to get a logger (also usefull for third party that would like to add debuging info). It might be part of another issue, but exploring this could help us make a choice (maybe).

Here is a case:

```python
# /third/part/starlite-websocket/entrypoint.py

from startlite.logging import get_logger

logger = get_logger(__name__)

logger.debug("initilializing listening server")
...
```

Under the hood, we need to ensure that get_logger return a logger according to user choice... which seems very tricky atm... I think we need to make a lazy call or something...

I also might have completly missed yhour point @Goldziher :D

Reply:
> I would like to avoid any configuration parsing to find a keyword. It looks like a needle in a haystack, it's hard to explain and seems to me very prone to unwanted behavior.

I completely agree here. 


> To untight Middleware from config, we need to provide a straightforward way to get a logger (also usefull for third party that would like to add debuging info). It might be part of another issue, but exploring this could help us make a choice (maybe).

I also think this is a good approach.

Reply:
I have made some tests and using a LoggingConfig parameter to choose which logging module user wants end up with complex and prone to bug code.

Here is a sample to illustrate (not tested yet):

```python
import sys

class LoggingConfig(BaseModel):
    """a wrapper for configuring any logging module that does not respect standard logging API."""

    logging_module: str = "logging"
        """Explicitly define the module name you want to use for logging."""

    def _get_logging_module(self) -> Callable:
            """Return instanciated module from logging_module configuration"""
            try:
                return sys.modules.get(
                    self.config.logging_module,
                    __import__(self.config.logging_module),  # if not already loaded
                )
            except ImportError as exc:
                raise ImportError(f"You need to manually install {self.module_name}") from exc

    def getLogger(self) -> Any:
        return self._get_logging_module().getLogger(self.config.middleware_logger_name)

    def configure(self) -> None:
        dict_config = {
            k: v for k, v in self.dict(exclude_none=True).items()
            if not k.startswith("middleware_") and k != "logging_module"
        }
        # calls config from logging_module
        config = getattr(self._get_logging_module(), "config")
        config.dictConfig(dict_config)
```

In another hand, making the same with inheritance is really more simple, using import statements top file (where it belongs) and explicitly saying which module to use seems to me way simpler:


```python
import logging
import logging.config
import picologging


class LoggingConfig(BaseModel):
    """a wrapper for configuring any logging module that does not respect standard logging API."""

    def getLogger(self) -> Any:
        return logging.getLogger(self.config.middleware_logger_name)

    def configure(self) -> None:
        dict_config = {
            k: v for k, v in self.dict(exclude_none=True).items()
            if not k.startswith("middleware_") and k != "logging_module"
        }
        # calls config from logging_module
        logging.config.dictConfig(dict_config)


class PiccologgingConfig(LoggingConfig):
    def getLogger(self) -> Any:
        return picologging.getLogger(self.config.middleware_logger_name)

    def configure(self) -> None:
        dict_config = {  # TODO move into dedicated method
            k: v for k, v in self.dict(exclude_none=True).items()
            if not k.startswith("middleware_") and k != "logging_module"
        }
        # calls config from logging_module
        picologging.config.dictConfig(dict_config)
```

Whatever our choice above, in the middleware (or whatever third party) we just have to call `self.config.getLogger()`.

In another issue, we should think on how to make this available through `from starlite import getLogger`.
We should also untight the "Convenience `pydantic` model for configuring logging" from app logging configuration.

I definitely vote for inheritance. 

@cofin do you agree ? Can I proceed ?

Reply:
this is not in scope for a middleware PR. Please see the new issue i created. Youre welcome to add a PR there.